src/v/kafka/client/client.cc

```diff
@@ -564,4 +564,39 @@ ss::future<kafka::fetch_response> client::consumer_fetch(
     });
 }
 
+ss::future<fetch_response> decompress_fetch_response(fetch_response resp) {
```

Comment:
This function does a lot of hidden work, which is somewhat forced by the fact that records are a `kafka::batch_reader`, which is based on `iobuf`.

The `iobuf` needs decoding, decompressing, and then reencoding.

The next user of the batch will then have to decode the `iobuf`, resulting in a wasteful encode/decode round.

I've contemplated a couple of strategies:

1. Changing records from a `kafka::batch_reader` to a `model::record_batch_reader` interface backed by a `kafka::batch_reader` impl. Decompression could then result in chaining of the `batch_reader` to a decompressor (so the work is done laziiy), hidden behind the interface. `model::record_batch_reader` is not a good type for encoding the response, but perhaps you can figure something out there.

2. Moving the decompress into `pandaproxy::json::rjson_serialize_impl<kafka::fetch_response>`, but that code isn't asynchronous.
    1. It looks like the underlying `compression::compressor::decompress` is actually synchronous (it probably shouldn't be), so it may be possible to expose a synchronous `storage::internal::decompress_batch`, at the cost of some tech debt.
    2. futurize `rjson_serialize_impl`


There are some other concerns along the way, but they should exist in another ticket which is related to streaming the data to the response body / not linearising the entire body.

Reply:
> The next user of the batch will then have to decode the iobuf, resulting in a wasteful encode/decode round.

Yeah

> It looks like the underlying compression::compressor::decompress is actually synchronous (it probably shouldn't be), so it may be possible to expose a synchronous storage::internal::decompress_batch, at the cost of some tech debt.

Precisely the reason I didn't just uncompress in the serialization functions.

> futurize rjson_serialize_impl

"easy"